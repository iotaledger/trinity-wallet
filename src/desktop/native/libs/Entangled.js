const fork = require('child_process').fork;
const path = require('path');
const EntangledNode = require('entangled-node');

let timeout = null;

/**
 * Spawn a child process and return result in async
 * @param {object} payload - Payload to send to the child process
 * @returns {Promise}
 */
const exec = (payload) => {
    return new Promise((resolve, reject) => {
        const child = fork(path.resolve(__dirname, 'Entangled.js'));

        const { job } = JSON.parse(payload);

        child.on('message', (message) => {
            resolve(message);

            clearTimeout(timeout);
            child.kill();
        });

        timeout = setTimeout(
            () => {
                reject(`Timeout: Entangled job: ${job}`);
                child.kill();
            },
            job === 'batchedPow' ? 180 * 1000 : 30 * 1000,
        );

        child.send(payload);
    });
};

/**
 * If module called as a child process, execute requested function and return response
 */
process.on('message', async (data) => {
    const payload = JSON.parse(data);

    if (payload.job === 'pow') {
        const pow = await EntangledNode.powTrytesFunc(payload.trytes, payload.mwm);
        process.send(pow);
    }

    if (payload.job === 'batchedPow') {
        const pow = await EntangledNode.powBundleFunc(
            payload.trytes,
            payload.trunkTransaction,
            payload.branchTransaction,
            payload.mwm,
        );
        process.send(pow);
    }

    if (payload.job === 'gen') {
        const address = await EntangledNode.genAddressTritsFunc(payload.seed, payload.index, payload.security);
        process.send(address);
    }
});

const Entangled = {
    powFn: async (trytes, mwm) => {
        return await exec(JSON.stringify({ job: 'pow', trytes, mwm }));
    },
    batchedPowFn: async (trytes, trunkTransaction, branchTransaction, mwm) => {
        return await exec(JSON.stringify({ job: 'batchedPow', trytes, trunkTransaction, branchTransaction, mwm }));
    },
    genFn: async (seed, index, security) => {
        return await exec(JSON.stringify({ job: 'gen', seed, index, security }));
    },
};

module.exports = Entangled;
