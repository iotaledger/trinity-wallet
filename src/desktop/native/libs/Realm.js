import Realm from 'realm';

// Don't report Realm JS analytics data
// @see https://github.com/realm/realm-js/blob/master/lib/submit-analytics.js#L28
process.env.REALM_DISABLE_ANALYTICS = 'true';

export default Realm;
