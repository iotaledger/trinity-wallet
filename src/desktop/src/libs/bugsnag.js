/* global Electron */
import React from 'react';
import bugsnag from '@bugsnag/js';
import bugsnagReact from '@bugsnag/plugin-react';

import settings from '../../package.json';

const bugsnagKey = process.env.BUGSNAG_KEY;

export const bugsnagClient = bugsnagKey.length
    ? bugsnag({
          apiKey: bugsnagKey,
          appVersion: settings.version,
          interactionBreadcrumbsEnabled: false,
          collectUserIp: false,
          user: { id: typeof Electron === 'object' ? Electron.getUuid() : null },
      })
    : {
          notify: (err) => {
              console.error(err); /* eslint-disable-line no-console */
          },
      };

if (bugsnagKey.length) {
    bugsnagClient.use(bugsnagReact, React);
}

export const ErrorBoundary = bugsnagKey.length ? bugsnagClient.getPlugin('react') : React.Fragment;
