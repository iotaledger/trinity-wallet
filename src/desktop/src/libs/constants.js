import { version } from '../../package.json';

/**
 * Use different account aliases for Relase Candidate versions
 */
export const __RC__ = version.toLowerCase().indexOf('rc') > 0 && process.env.NODE_ENV !== 'test' ? '-rc' : '';

/**
 * Main account encryption key alias
 */
export const ALIAS_MAIN = `Trinity${__RC__}`;

/**
 * Seed account encryption key alias
 */
export const ALIAS_ACCOUNT = `account${__RC__}`;

/**
 * Realm database encryption key alias
 */
export const ALIAS_REALM = `realm_enc_key${__RC__}`;

/**
 * Legacy credentials key alias
 */
export const ALIAS_LEGACY = `moonpay_credentials${__RC__}`;
