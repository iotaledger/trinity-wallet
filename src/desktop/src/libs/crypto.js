/* global Electron */
import { MAX_SEED_LENGTH } from 'libs/iota/utils';

// Prefix for seed account titles stored in the vault
const ACC_PREFIX = 'account';

// Title of main vault entry  containing 2fa keys
export const ACC_MAIN = 'Trinity';
// Maximum allowed account title
export const MAX_ACC_LENGTH = 250;

/**
 * Create random byte array
 * @param {Number} Length - Random number array length.
 * @param {Number} Max -  Random byte max range
 * @returns {Array} Random number array
 */
function randomBytes(size, max) {
    if (size !== parseInt(size, 10) || size < 0) {
        return false;
    }

    const rawBytes = new Uint8Array(size);

    const bytes = global.crypto.getRandomValues(rawBytes);

    for (let i = 0; i < bytes.length; i++) {
        while (bytes[i] >= 256 - (256 % max)) {
            bytes[i] = randomBytes(1, max)[0];
        }
    }

    return Array.from(bytes);
}

/**
 * Create random seed
 * @param {Number} length - Random seed length
 * @returns {Array} Random byte array seed
 */
export const createRandomSeed = (length = MAX_SEED_LENGTH) => {
    return randomBytes(length, 27);
};

/**
 * Encrypt plain text
 * @param {Any} Content - Content to encrypt
 * @param {String} Password - Plain text password for encryption
 * @returns {String} Ecnrypted initialization vector + content
 */
const encrypt = async (contentPlain, passwordPlain) => {
    const content = new TextEncoder().encode(JSON.stringify(contentPlain));

    const password = new TextEncoder().encode(passwordPlain);
    const passwordhash = await crypto.subtle.digest('SHA-256', password);

    const iv = crypto.getRandomValues(new Uint8Array(12));

    const algorithm = { name: 'AES-GCM', iv: iv };

    const key = await crypto.subtle.importKey('raw', passwordhash, algorithm, false, ['encrypt']);

    const cipherBuffer = await crypto.subtle.encrypt(algorithm, key, content);
    const cipherArray = new Uint8Array(cipherBuffer);
    const cipherHex = cipherArray.toString();

    const ivHex = iv.toString();

    return `${ivHex}|${cipherHex}`;
};

/**
 * Decrypt cyphertext
 * @param {String} Content - Encrypted initialization vector + content
 * @param {String} Password - Plain text password for decryption
 * @returns {Any} Derypted content
 */
const decrypt = async (cipherText, passwordPlain) => {
    const cipherParts = cipherText.split('|');

    if (cipherParts.length !== 2 || typeof passwordPlain !== 'string' || !passwordPlain.length) {
        throw new Error('Wrong password');
    }
    try {
        const password = new TextEncoder().encode(passwordPlain);
        const passwordHash = await crypto.subtle.digest('SHA-256', password);

        const ivArray = cipherParts[0].split(',');
        const iv = Uint8Array.from(ivArray);

        const algorithm = { name: 'AES-GCM', iv: iv };

        const key = await crypto.subtle.importKey('raw', passwordHash, algorithm, false, ['decrypt']);

        const cipherArray = cipherParts[1].split(',');
        const cipher = Uint8Array.from(cipherArray);

        const plainBuffer = await crypto.subtle.decrypt(algorithm, key, cipher);
        const plainText = new TextDecoder().decode(plainBuffer);

        return JSON.parse(plainText);
    } catch (err) {
        throw new Error('Wrong password');
    }
};

/**
 * Check for valid vault password
 * @param {String} Password - Plain text password for decryption
 * @returns {String} Two factor authentication key
 */
export const vaultAuth = async (password) => {
    const vault = await Electron.readKeychain(ACC_MAIN);
    if (!vault) {
        throw new Error('Local storage not available');
    }
    try {
        const decryptedVault = await decrypt(vault, password);
        if (decryptedVault.twoFaKey) {
            return decryptedVault.twoFaKey;
        }
        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Clear the vault
 * @param {String} Password - Plain text password for decryption
 * @param {Boolean} Reset - Should the vault be reset without authentication
 * @returns {Boolean} True if vault cleared
 */
export const clearVault = async (password, reset) => {
    if (!reset) {
        await vaultAuth(password);
    }

    const vault = await Electron.listKeychain();

    const accounts = Object.keys(vault);

    for (let i = 0; i < accounts.length; i++) {
        await Electron.removeKeychain(vault[i].account);
    }

    return true;
};

/**
 * Update vault password
 * @param {String} PasswordCurrent - Current plain text password
 * @param {String} PasswordNew - New plain text password
 * @returns {Boolean} True if password updated
 */
export const updatePassword = async (passwordCurrent, passwordNew) => {
    const vault = await Electron.listKeychain();
    if (!vault) {
        throw new Error('Local storage not available');
    }
    try {
        const accounts = Object.keys(vault);

        if (!accounts.length) {
            return true;
        }

        for (let i = 0; i < accounts.length; i++) {
            const account = vault[i];

            const decryptedVault = await decrypt(account.password, passwordCurrent);
            const encryptedVault = await encrypt(decryptedVault, passwordNew);

            Electron.setKeychain(account.account, encryptedVault);
        }

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Get seed from keychain
 * @param {String} Password - Plain text password for decryption
 * @param {String} SeedName - Seed name to retreive from keychain
 * @param {Boolean} PlainText - Should the seed be returned in plain ACII text
 * @returns {Array} Derypted seed
 */
export const getSeed = async (password, seedName, plainText) => {
    const seedNameHash = await hashSeedName(seedName);

    const vault = await Electron.readKeychain(seedNameHash);
    if (!vault) {
        throw new Error('Incorrect seed name');
    }
    try {
        const decryptedVault = await decrypt(vault, password);
        return plainText ? seedToHex(decryptedVault) : decryptedVault;
    } catch (err) {
        throw err;
    }
};

/**
 * Save seed to keychain
 * @param {String} Password - Plain text password for encryption
 * @param {String} SeedName - Seed name to set to keychain
 * @param {Array} Seed - Seed array
 * @returns {Boolean} True if seed is saved to keychain
 */
export const setSeed = async (password, seedName, seed) => {
    try {
        const seedNameHash = await hashSeedName(seedName);
        const vault = await encrypt(Array.from(seed), password);

        Electron.setKeychain(seedNameHash, vault);

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Set Two-Factor authentication key
 * @param {String} Password - Plain text password for decryption
 * @param {String} Key - Two-factor authentication key
 * @returns {Boolean} True if Two-Factor key is set
 */
export const setTwoFA = async (password, key) => {
    try {
        const vault = await Electron.readKeychain(ACC_MAIN);
        const decryptedVault = vault === null ? {} : await decrypt(vault, password);

        if (key) {
            decryptedVault.twoFaKey = key;
        } else {
            delete decryptedVault.twoFaKey;
        }

        const updatedVault = await encrypt(decryptedVault, password);

        Electron.setKeychain('Trinity', updatedVault);

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Remove seed from keychain
 * @param {String} Password - Plain text password for decryption
 * @param {String} SeedName - Seed name to remove from keychain
 * @returns {Boolean} True seed is removed
 */
export const removeSeed = async (password, seedName) => {
    try {
        const seedNameHash = await hashSeedName(seedName);

        const isRemoved = await Electron.removeKeychain(seedNameHash);
        if (!isRemoved) {
            throw new Error('Incorrect seed name');
        }

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Rename seed in keychain
 * @param {String} Password - Plain text password for decryption
 * @param {String} OldName - Current seed name
 * @param {String} NewName - New seed name
 * @returns {Boolean} True if seed is renamed
 */
export const renameSeed = async (password, seedName, newSeedName) => {
    const seedNameHash = await hashSeedName(seedName);
    const newNameHash = await hashSeedName(newSeedName);

    const vault = await Electron.readKeychain(seedNameHash);

    if (!vault) {
        throw new Error('Incorrect seed name');
    }

    try {
        await decrypt(vault, password);
        Electron.removeKeychain(seedNameHash);
        Electron.setKeychain(newNameHash, vault);

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Unique seed check
 * @param {String} Password - Plain text password for decryption
 * @param {Array} Seed - Seed to check
 * @returns {Boolean} True if seed is unique
 */
export const uniqueSeed = async (password, seed) => {
    const vault = await Electron.listKeychain();
    if (!vault) {
        throw new Error('Local storage not available');
    }
    try {
        const accounts = vault.filter((acc) => acc.account !== ACC_MAIN);

        for (let i = 0; i < accounts.length; i++) {
            const account = vault[i];
            const vaultSeed = await decrypt(account.password, password);
            if (vaultSeed.length === seed.length && seed.every((v, x) => v % 27 === vaultSeed[x] % 27)) {
                return false;
            }
        }

        return true;
    } catch (err) {
        throw err;
    }
};

/**
 * Hash text using SHA-256
 * @param {String} Password - plain text to hash
 * @returns {String} SHA-256 hash
 */
export const sha256 = async (inputPlain) => {
    if (typeof inputPlain !== 'string' || inputPlain.length < 1) {
        return false;
    }
    const input = new TextEncoder().encode(inputPlain);
    const hash = await crypto.subtle.digest('SHA-256', input);
    const plainHash = bufferToHex(hash);

    return plainHash;
};

/**
 * Hash seed name using SHA-256
 * @param {String} SeedName - plain text to hash
 * @returns {String} SHA-256 hash
 */
const hashSeedName = async (seedName) => {
    const prefixName = `${ACC_PREFIX}-${seedName}`;
    const hash = await sha256(prefixName);
    return hash;
};

/**
 * Check for a valid activation code
 * @param {String} code - Target activation code
 * @param {String} uuid - UUID of the machine
 * @returns {Boolean} True if activation code correct
 */
export const checkActivationCode = async (code, uuid) => {
    const key = 'LURGzCPEHqhjvYLwAJXRv5Fc';
    const hash = await sha256(key + uuid);
    return code === hash;
};

/**
 * Convert byte seed array to string
 * @param {Array} seed - Target seed array
 * @returns {String} Plain text seed string
 */
const seedToHex = (bytes) => {
    return Array.from(bytes)
        .map((byte) => byteToChar(byte % 27))
        .join('');
};

/**
 * Convert single character byte to string
 * @param {Number} byte - Input byte
 * @returns {String} Output character
 */
export const byteToChar = (byte) => {
    return '9ABCDEFGHIJKLMNOPQRSTUVWXYZ'.charAt(byte % 27);
};

/**
 * Convert buffer to plain text
 * @param {Buffer} - Input buffer
 * @returns {String} Output string
 */
const bufferToHex = (buffer) => {
    const hexCodes = [];
    const view = new DataView(buffer);
    for (let i = 0; i < view.byteLength; i += 4) {
        const value = view.getUint32(i);
        const stringValue = value.toString(16);
        const padding = '00000000';
        const paddedValue = (padding + stringValue).slice(-padding.length);
        hexCodes.push(paddedValue);
    }

    return hexCodes.join('');
};
