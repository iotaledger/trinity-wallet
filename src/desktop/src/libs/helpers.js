/**
 * Capitalizes a string
 * @param {String} input - The target string
 * @returns {String} Capitalized string
 */
export const capitalize = (input) => {
    return typeof input === 'string' && input.length > 0 ? input[0].toUpperCase() + input.substr(1).toLowerCase() : '';
};

/**
 * Trim a string to certain size
 * @param {String} input - The target string
 * @param {Number} length - The target length
 * @returns {String} Trimmed string
 */
export const shorten = (input, length) => {
    return typeof input === 'string' && input.length > length ? input.substr(0, length - 1) + 'â€¦' : input;
};
