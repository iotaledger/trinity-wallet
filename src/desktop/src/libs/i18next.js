import i18next from 'i18next';

const I18N_DEFAULT_LANGUAGE = 'en';

export default i18next
    .init({
        // lng: I18N_DEFAULT_LANGUAGE,
        fallbackLng: I18N_DEFAULT_LANGUAGE,
        fallbackNS: 'Common',
        parseMissingKeyHandler: (missing) => `MISSING TRANSLATION: ${missing}`,
        resources: {
            // de: { 'Common': require('../../../shared/locales/de/translation.json') },
            // en: { 'Common': require('../../../shared/locales/en/translation.json') },
            // 'es-ES': { 'Common': require('../../../shared/locales/es-ES/translation.json') },
            // fr: { 'Common': require('../../../shared/locales/fr/translation.json') },
            en: require('../../../shared/locales/en/translation.json'),
            es: require('../../../shared/locales/es/translation.json'),
        },
    });
