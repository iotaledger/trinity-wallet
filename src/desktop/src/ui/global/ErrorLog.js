/* global Electron */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import { format } from 'date-fns';

import { clearLog } from 'actions/alerts';

import Modal from 'ui/components/modal/Modal';
import Button from 'ui/components/Button';
import Scrollbar from 'ui/components/Scrollbar';

import css from './about.scss';

/**
 * Error log window component
 */
class ErrorLog extends React.PureComponent {
    static propTypes = {
        /** @ignore */
        log: PropTypes.array.isRequired,
        /** @ignore */
        clearLog: PropTypes.func.isRequired,
        /** @ignore */
        t: PropTypes.func.isRequired,
    };

    state = {
        visible: false,
    };

    componentDidMount() {
        this.onMenuToggle = this.menuToggle.bind(this);
        Electron.onEvent('menu', this.onMenuToggle);
    }

    componentWillUnmount() {
        Electron.removeEvent('menu', this.onMenuToggle);
    }

    /**
     * Proxy native menu triggers to an action
     * @param {string} Item - Triggered menu item
     */
    menuToggle(item) {
        this.setState({
            visible: item === 'errorlog',
        });
    }

    render() {
        const { clearLog, log, t } = this.props;
        const { visible } = this.state;

        return (
            <Modal variant="global" isOpen={visible} onClose={() => this.setState({ visible: false })}>
                <section className={css.about}>
                    <h5>{t('notificationLog:errorLog')}</h5>
                    <article>
                        <Scrollbar>
                            <ul className={css.log}>
                                {log
                                    .slice(0)
                                    .reverse()
                                    .map((entry, index) => {
                                        return (
                                            <li key={`log-${index}`}>
                                                <strong>{format(entry.time, 'DD.MM.YY HH:mm')}</strong> {entry.error}
                                            </li>
                                        );
                                    })}
                            </ul>
                        </Scrollbar>
                    </article>
                    <footer>
                        <Button onClick={() => clearLog()} variant="dark">
                            {t('clear')}
                        </Button>
                        <Button onClick={() => this.setState({ visible: false })} variant="dark">
                            {t('close')}
                        </Button>
                    </footer>
                </section>
            </Modal>
        );
    }
}

const mapStateToProps = (state) => ({
    log: state.alerts.notificationLog,
});

const mapDispatchToProps = {
    clearLog,
};

export default connect(mapStateToProps, mapDispatchToProps)(translate()(ErrorLog));
