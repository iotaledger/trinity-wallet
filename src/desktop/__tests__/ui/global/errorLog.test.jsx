import React from 'react';
import { shallow } from 'enzyme';

import { ErrorLogComponent as ErrorLog } from 'ui/global/ErrorLog';

import * as dateLib from 'libs/date';
dateLib.formatTimeAs = { dayMonthYearHoursMinutes: () => 'DD/MM/YYYY HH:II' };

const props = {
    log: Array(10).fill({
        date: 1548842449,
        error: 'Foo',
    }),
    clearLog: jest.fn(),
    t: (str) => str,
};

global.Electron = {
    onEvent: function(_event, e) {
        this.menuEvent = e;
    },
    removeEvent: () => {},
    menuEvent: null,
};

describe('ErrorLog component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<ErrorLog {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Render error entries', () => {
        const wrapper = shallow(<ErrorLog {...props} />);

        expect(wrapper.find('li')).toHaveLength(10);
    });

    test('Set visible on menu event', () => {
        const wrapper = shallow(<ErrorLog {...props} />);

        Electron.menuEvent('errorlog');
        expect(wrapper.state('visible')).toBeTruthy();
    });
});
