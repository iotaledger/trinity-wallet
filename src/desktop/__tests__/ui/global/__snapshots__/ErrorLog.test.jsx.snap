// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorLog component Render the component 1`] = `
<Modal
  isOpen={false}
  onClose={[Function]}
  variant="global"
>
  <section
    className="about"
  >
    <h5>
      notificationLog:errorLog
    </h5>
    <article>
      <Scrollbar>
        <ul
          className="log"
        >
          <li
            key="log-0"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-1"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-2"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-3"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-4"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-5"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-6"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-7"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-8"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
          <li
            key="log-9"
          >
            <strong>
              DD/MM/YYYY HH:II
            </strong>
             
            Foo
          </li>
        </ul>
      </Scrollbar>
    </article>
    <footer>
      <Button
        onClick={[Function]}
        variant="dark"
      >
        clear
      </Button>
      <Button
        onClick={[Function]}
        variant="dark"
      >
        close
      </Button>
    </footer>
  </section>
</Modal>
`;
